Sidekiq refactor

matsimitsu opened the pull request:
Fork of #347 with a green testsuite

rmachielse commented on lib/appsignal/hooks/sidekiq.rb:
```diff
@@ -14,23 +14,21 @@ class args retried_at failed_at
+        job = ::Sidekiq::Job.new(item)
+
+        params = Appsignal::Utils::ParamsSanitizer.sanitize(
+          job.display_args,
```
Small heads up, I noticed that the original args were mutated by the sanitizer, causing `display_class` to fail later. Not sure if it was actually the sanitizer causing it. It was consistent so if you don't run into it with a real sidekiq job it's probably no longer a problem.

tombruijn commented on lib/appsignal/hooks/sidekiq.rb:
```diff
@@ -53,6 +51,7 @@ def dependencies_present?
       end
 
       def install
+        require "sidekiq/api"
```
Wasn't sure we decided to go with this yet? It loads something that's not loaded by default and I'm not sure we'd want that.

matsimitsu replied:
No decision has been made yet, but I wanted to at least have a green build to see if it worked :)

tombruijn replied:
Mike Perham's response on using this worries me still:

> note that this method was written for Web UI usage, the output could change in the future.

If we rely on this internal class our integration might break with future versions of Sidekiq if they change things around.

jeffkreeftmeijer replied:
If I understand correctly, we’re requiring “sidekiq/api” here specifically for the [`Sidekiq::Job#display_class`](https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/api.rb#L316-L334) and [`Sidekiq::Job#sidplay_args`](https://github.com/mperham/sidekiq/blob/master/lib/sidekiq/api.rb#L336-L358) methods, which produce “human-friendly” strings.

Requiring this file makes sense, because that means we won't have to write these functions ourselves. However, they're from an internal API which could potentially change without notice in future Sidekiq versions. This could break our instrumentation, or --because the format can change without breaking our implementation-- it can change how events are grouped in AppSignal.

Another option would be to copy these methods into the project to gain absolute control over what matches, and what we send to AppSignal. We'd need to move these two functions (including their tests) to our code base, and maintain them from now on. This approach could also break on a new version of Sidekiq, but at least we're sure that the format won't change.

rmachielse replied:
Hi jeff!

Of course this is a very valid concern.
My reasoning however was that the underlying "raw data" is as much an internal matter as these methods are.

Next to that, Mike Perham also said:

> It's unlikely to change but I can't control AJ so it's hard to say.

So the question is, what is more likely to change? The methods' name/location/output, or the underlying data format? Copying the methods will prevent problems in the first case, while it will introduce new ones in case of the second.

matsimitsu replied:
> Requiring this file makes sense, because that means we won't have to write these functions ourselves. However, they're from an internal API which could potentially change without notice in future Sidekiq versions. This could break our instrumentation, or --because the format can change without breaking our implementation-- it can change how events are grouped in AppSignal.

The thing with extracting it is that we only move the problem around, because when the Sidekiq argument implementation changes our own extracted methods are also broken. 

So regardless of what option we choose, if Sidekiq changes, our integration is broken. My argument for using this API is that it will be tested a lot better then our own implementation, as Sidekiq's focus is 100% on Sidekiq, while our extracted implementation is just a small detail in our gem.

jeffkreeftmeijer replied:
None of these options make it less likely for failure because we can’t control the input. Vendoring these methods in is a suggestion to have control over the output, at least. However, as @rmachielse said (hi Richard! :wave:), fiddling with the data is as likely to break as calling an internal method. Also, I trust Mike to maintain an internal Sidekiq method better than we could, and I trust our customers to let us know if something like this breaks. Let's ship this as is and revisit this issue later, if we need to.

tombruijn replied:
One thing we could do to prevent a catastrophic failure is to wrap the call to the sidekiq method in a begin-rescue block?

something like:

```ruby
        job_action_name =
          begin
            action_name = job.display_class
            if action_name =~ /[\.#]+/
              action_name
            else
              # Add #perform as default method name for job names without a method
              # name
              "#{action_name}#perform"
            end
          rescue NoMethodError => e
            Appsignal.logger.error("Problem parsing the Sidekiq job name: #{e}")
          end
```

matsimitsu replied:
sounds like a good idea :)

tombruijn commented on spec/lib/appsignal/hooks/sidekiq_spec.rb:
```diff
@@ -1,145 +1,69 @@
 describe Appsignal::Hooks::SidekiqPlugin do
```
I can't find a test in this spec file that tests this new scenario?

A display class that is `Model.delay.method` or something

tombruijn commented:
Moved to "Waiting" until we have a moment to discuss what solution we want to implement.

rmachielse commented:
@tombruijn @matsimitsu any idea when this will be continued? We are quite desperate for this, as we're currently in the dark about our background jobs' performance and exception rate

tombruijn commented:
We've decided to go with this solution. @matsimitsu will pick this up again and see if he can fix the failing specs.

@rmachielse our sanitizer should not mutate any params it's given, but we'll have another look just in case.

rmachielse commented:
@matsimitsu @tombruijn awesome guys, thanks a lot 👍

matsimitsu commented:
@rmachielse I'm not able to write a failing spec for the param sanitizer. Do you have a bit more information about what you were observing when using the sanitizer by any chance?

rmachielse commented:
@matsimitsu I think the `display_class` method started to throw an exception. But this was only with real sidekiq, not with the mock in the specs. I can try to reproduce if you want

matsimitsu commented:
@rmachielse that'd be great!

rmachielse commented:
@matsimitsu I was able to reproduce the problem, but it looks like it was caused by some jobs in my local redis that came from a different project. After clearing them out everything worked great with this branch. My bad, looks like everything is fine!

matsimitsu commented:
@rmachielse Awesome! I'll start the process of getting this to master :)

rmachielse commented:
Although, wait a minute, I might have it again. Sorry, looking into it now

rmachielse commented:
@matsimitsu I do know how to prevent it at least. If you call `job.display_class` before sanitizing, the value is cached (that's why I thought it didn't happen anymore). Otherwise, I'm seeing this error:

```ruby
16:33:51 worker.1 | 2017-10-02T14:33:51.530Z 95407 TID-oxiuzi0c0 WARN: NoMethodError: undefined method `[]' for nil:NilClass
16:33:51 worker.1 | 2017-10-02T14:33:51.530Z 95407 TID-oxiuzi0c0 WARN: /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/api.rb:327:in `display_class'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/bundler/gems/appsignal-ruby-7ae15b66af7c/lib/appsignal/hooks/sidekiq.rb:24:in `call'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/middleware/chain.rb:130:in `block in invoke'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/middleware/chain.rb:133:in `invoke'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:169:in `block in process'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:141:in `block (6 levels) in dispatch'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/job_retry.rb:97:in `local'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:140:in `block (5 levels) in dispatch'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/rails.rb:42:in `block in call'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/activesupport-5.1.4/lib/active_support/execution_wrapper.rb:85:in `wrap'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/activesupport-5.1.4/lib/active_support/reloader.rb:68:in `block in wrap'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/activesupport-5.1.4/lib/active_support/execution_wrapper.rb:85:in `wrap'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/activesupport-5.1.4/lib/active_support/reloader.rb:67:in `wrap'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/rails.rb:41:in `call'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:136:in `block (4 levels) in dispatch'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:215:in `stats'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:131:in `block (3 levels) in dispatch'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/job_logger.rb:7:in `call'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:130:in `block (2 levels) in dispatch'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/job_retry.rb:72:in `global'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:129:in `block in dispatch'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/logging.rb:44:in `with_context'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/logging.rb:38:in `with_job_hash_context'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:128:in `dispatch'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:168:in `process'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:85:in `process_one'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/processor.rb:73:in `run'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/util.rb:16:in `watchdog'
16:33:51 worker.1 | /usr/local/rvm/gems/ruby-2.4.2/gems/sidekiq-5.0.4/lib/sidekiq/util.rb:25:in `block in safe_thread
```

rmachielse commented:
@matsimitsu I think the sanitizer does actually mutate the `args` somehow. If I use `puts job.instance_variable_get('@args')` before and after sanitizing, I get a different result

rmachielse commented:
@matsimitsu next to that, it only seems to happen with mailers

rmachielse commented:
@matsimitsu ok, latest news, it seems to be `job.display_args` itself that is causing the problem. Putting `display_class` in a variable before sanitizing instead of after solves the problem. However the real problem is a bug in sidekiq in that case

rmachielse commented:
@matsimitsu I have created an issue on sidekiq as I think it can be fixed there: mperham/sidekiq#3620. Do you want to wait for that?

rmachielse commented:
@matsimitsu the fix on sidekiq is now merged and released to `v5.0.5`. Unfortunately I also discovered that the `display_class` method sometimes returns `Class.method` and sometimes `Class#method`. I attempt to fix that in mperham/sidekiq#3623. However if appsignal wants to be backwords compatible with older releases:

- `display_class` has to be called before `display_args`
- the class should be split like this: `job.display_class.split(/\.|#/, 2)`

matsimitsu commented:
@rmachielse Thanks for your help! I've changed the code and added a spec. We'll run this branch on our staging today and if all goes well we can get this to master :)

rmachielse commented:
@matsimitsu thanks, looking forward to it!
I verified and everything is working great now

tombruijn commented on gemfiles/sidekiq.gemfile:
```diff
@@ -0,0 +1,6 @@
+source 'https://rubygems.org'
+
+gem 'rails', '~> 5.1.0'
```
is Rails really required for the sidekiq specs?

rmachielse commented:
@tombruijn true, but sidekiq doesn't seem to have an alternative itself.
One alternative would be to parse the arguments' YAML directly, but I'm not sure if it is always the case that the first argument is the class, the second the method and the others it's arguments

tombruijn commented on lib/appsignal/hooks/sidekiq.rb:
```diff
@@ -14,23 +14,24 @@ class args retried_at failed_at
-          :class       => item["wrapped"] || item["class"],
-          :method      => "perform",
+          :class       => display_class,
+          :method      => display_method || "perform",
```
To show the difference between a class and instance method here for the Sidekiq delayed extension we might need to drop `Appsignal.monitor_transaction` usage and manually create a transaction.

The implementation in the [`Transaction#set_http_or_background_action` method](https://github.com/appsignal/appsignal-ruby/blob/eb421a537b8ebaac0d66fc20f002276e9998b489/lib/appsignal/transaction.rb#L225) always assumes a instance method.

Follow up from my reply here: https://github.com/mperham/sidekiq/pull/3623#pullrequestreview-66724350

tombruijn commented:
I've [added tests](https://github.com/appsignal/appsignal-examples/commit/aa3db6cbf6715459ea03d66c951b8dd5fc104eeb) to the sidekiq example app in our examples repo. That should allow for easier testing with real data.

rmachielse commented:
@matsimitsu @tombruijn what is the status of this now? Anything I can do?

tombruijn commented:
Hi @rmachielse ,

Robert is a little busy curretnly and I'm juggling support and finishing up work on #351 (still).

I'm going to see if I can make the changes I suggested this week, otherwise it will be first thing next Monday.

Are you able to use your original change until this has been resolved?

tombruijn commented:
@matsimitsu would be nice if we could run this on staging and see if all this works :)

rmachielse commented:
@tombruijn nice! I tried the new version and everything is working fine.
I'm indeed seeing both `.` and `#` notation now in appsignal

tombruijn commented:
@rmachielse great! Thanks for reporting back. Now we just need some reviews on the code before we can merge this.

tombruijn commented:
@matsimitsu we have that problem again where this is your PR and you can't approve it on GitHub, but can you review and give a 👍  if you approve?

tombruijn commented:
@matsimitsu @jeffkreeftmeijer added some defensive programming around the calls to `Sidekiq::Job` instances in commit: https://github.com/appsignal/appsignal-ruby/pull/348/commits/6402730d298b116130588e845e4f68b662e6fc9f
Do you approve ?

tombruijn commented:
Oh and I tested this with ActiveJob (example added to https://github.com/appsignal/appsignal-examples/tree/sidekiq)
and Sidekiq v3.5.4, to make sure we support a couple versions back.

matsimitsu commented:
@tombruijn approved!

tombruijn commented:
Noticed one tiny little problem with a `call_and_log_on_error` call, now fixed and pushed.

Once the build is green I'll merge the PR and release Ruby gem 2.3.7

tombruijn commented:
@rmachielse AppSignal for Ruby 2.3.7 with this fixed has been released. Let us know if you experience any problems with this version.

rmachielse commented:
@tombruijn thanks a lot! We've deployed it to our staging environment and everything seems to work perfectly

tombruijn commented:
Heads-up @rmachielse, we've reverted this change in the AppSignal Ruby gem v2.4.0 release. It's breaking backwards compatibility for older Sidekiq versions and we're experiencing problems with manually inserted jobs. I've created an issue for this in #360

rmachielse commented:
@tombruijn that's unfortunate. The last v2 release is from february 2014. How do you want to continue?

tombruijn commented:
we've reverted the change in v2.4.0 and we'll fix it in a new 2.4.x release. I think we're just going to have to copy the action naming logic sidekiq uses in our own setup.

tombruijn commented:
@rmachielse We've redone this change in PR #362. If you have time, can you try it out and let me know if it still works as expected?

rmachielse commented:
@tombruijn I've tried and everything seems to work perfectly! Thanks!

tombruijn commented:
@rmachielse that's great to hear! We'll try and squeeze in a point release for 2.4.1 sometime soon

tombruijn commented:
@rmachielse this is now released in AppSignal gem v2.4.1 🎉

rmachielse commented:
@tombruijn thanks! 🎉
